src/kafka-util/src/admin.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Helpers for working with Kafka's admin API.
+
+use std::error::Error;
+use std::fmt;
+use std::iter;
+use std::time::Duration;
+
+use rdkafka::admin::{AdminClient, AdminOptions, NewTopic};
+use rdkafka::client::ClientContext;
+use rdkafka::error::{KafkaError, RDKafkaError};
+
+use ore::collections::CollectionExt;
+use ore::retry;
+
+/// Creates a Kafka topic and waits for it to be reported in the broker
+/// metadata.
+///
+/// This function is a wrapper around [`AdminClient::create_topics`] that
+/// attempts to ensure the topic creation has propagated throughout the Kafka
+/// cluster before returning. Kafka topic creation is asynchronous, so
+/// attempting to consume from or produce to a topic immediately after its
+/// creation can result in "unknown topic" errors.
+///
+/// This function does not return successfully unless it can find the metadata
+/// for the newly-created topic in a call to [`Client::fetch_metadata`] and
+/// verify that the metadata reports the topic has the number of partitions
+/// requested in `new_topic`. Empirically, this seems to be the condition that
+/// guarantees that future attempts to consume from or produce to the topic will
+/// succeed.
+pub async fn create_topic<'a, C>(
```

Comment:
so just a thought - do you think this ought to be upstreamed into rust-rdkafka?

Reply:
In principle, yes, but anything that goes in rust-rdkafka has a much higher maintenance burden because it has to work for any user of the library, not just us. I'd be willing to believe that there are still ways this topic syncing goes wrong; it's ok for us because we only need it to work in tests, but it goes into rust-rdkafka it's liable to get used in more production-like environments.

To be honest it's really silly that the create topic request can return before the topic is actually usable. There's all sorts of bugs filed against Kafka/librdkafka about this:

  * https://github.com/confluentinc/confluent-kafka-python/issues/524#issuecomment-456783176
  * https://github.com/confluentinc/confluent-kafka-python/issues/856
  * https://github.com/apache/kafka/pull/7343

tl;dr if someone else wanted to maintain this stuff in rust-rdkafka, for sure, but while it's just me I'd rather not open that can of worms.